swarm/storage/localstore/subscriptions.go

```diff
@@ -0,0 +1,261 @@
+// to control and get information about subscription state.
+type Subscription struct {
+	// Chunks is the read-only channel that provides stream of chunks.
+	// This is the subscription main purpose.
+	Chunks <-chan storage.Chunk
+
+	// subscribe to set of keys only with this prefix
+	prefix []byte
+	// signals subscription to gracefully stop
+	stopChan chan struct{}
+	// protects stopChan form multiple closing
+	stopOnce sync.Once
+	// provides information if subscription is done
+	doneChan chan struct{}
+	// trigger signals a new index iteration
+	// when index receives new items
+	trigger chan struct{}
+	// an error from the subscription, if any
+	err error
+	// protects err field
+	mu sync.RWMutex
+}
+
+// Done returns a read-only channel that will be closed
+// when the subscription is stopped or encountered an error.
+func (s *Subscription) Done() <-chan struct{} {
+	return s.doneChan
+}
+
+// Err returns an error that subscription encountered.
+// It should be usually called after the Done is read from.
+// It is safe to call this function multiple times.
+func (s *Subscription) Err() (err error) {
+	s.mu.RLock()
+	err = s.err
+	s.mu.RUnlock()
+	return err
+}
+
+// Stop terminates the subscription without any error.
+// It is safe to call this function multiple times.
+func (s *Subscription) Stop() {
+	s.stopOnce.Do(func() {
+		close(s.stopChan)
+	})
+}
+
+// feed is a collection of Chunks subscriptions of order given
+// by sort index and Chunk data provided by data index.
+// It provides methods to create, trigger and remove subscriptions.
+// It is the internal core component for push and pull
+// index subscriptions.
+type feed struct {
+	// index on which keys the order of Chunks will be
+	// provided by subscriptions
+	sortIndex shed.Index
+	// index that contains chunk data
+	dataIndex shed.Index
+	// collection fo subscriptions on this feed
+	subscriptions []*Subscription
+	// protects subscriptions slice
+	mu sync.Mutex
+	// closed when subscription is closed
+	closeChan chan struct{}
+	// protects closeChan form multiple closing
+	closeOnce sync.Once
+}
+
+// newFeed creates a new feed with from sort and data indexes.
+// Sort index provides ordering of Chunks and data index
+// provides Chunk data.
+func newFeed(sortIndex, dataIndex shed.Index) (f *feed) {
+	return &feed{
+		sortIndex:     sortIndex,
+		dataIndex:     dataIndex,
+		subscriptions: make([]*Subscription, 0),
+		closeChan:     make(chan struct{}),
+	}
+}
+
+// subscribe creates a new subscription on the feed.
+// It creates a new goroutine which will iterate over existing sort index keys
+// and creates new iterators when trigger method is called.
+func (f *feed) subscribe(ctx context.Context, prefix []byte) (s *Subscription, err error) {
+	// prevent new subscription after the feed is closed
+	select {
+	case <-f.closeChan:
+		return nil, ErrSubscriptionFeedClosed
+	default:
+	}
+	chunks := make(chan storage.Chunk)
+	s = &Subscription{
+		Chunks:   chunks,
```

Comment:
The iterator should just give hashes back, no need for chunk data
Shall we reiterate on this? I am starting to feel that instead of subscribe we better use an iterator 
we can hide all the internal bookkeeping from the clients

Reply:
Subscriptions are now changed to be for hashes instead chunks. I am open for discussion what is actually needed.